// drei/prototyping/Loader.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _extends = require('@babel/runtime/helpers/extends');
var React = require('react');
require('three');
require('zustand');
var loaders_useProgress_cjs = require('../loaders/useProgress.cjs.js');
var web = require('@react-spring/web');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _extends__default = /*#__PURE__*/_interopDefaultLegacy(_extends);
var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

var styles = {
  container: {
    position: 'absolute',
    top: 0,
    left: 0,
    width: '100%',
    height: '100%',
    background: '#171717',
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    zIndex: 1000
  },
  inner: {
    width: 100,
    height: 3,
    background: '#272727',
    textAlign: 'center'
  },
  bar: {
    height: 3,
    width: '100%',
    background: 'white',
    transformOrigin: 'left center'
  },
  data: {
    display: 'inline-block',
    position: 'relative',
    fontVariantNumeric: 'tabular-nums',
    marginTop: '0.8em',
    color: '#f0f0f0',
    fontSize: '0.6em',
    fontFamily: "-apple-system, BlinkMacSystemFont, \"Inter\", \"Segoe UI\", \"Helvetica Neue\", Helvetica, Arial, Roboto, Ubuntu, sans-serif, \"Apple Color Emoji\", \"Segoe UI Emoji\", \"Segoe UI Symbol\"",
    whiteSpace: 'nowrap'
  }
};
function Loader(_ref) {
  var containerStyles = _ref.containerStyles,
      innerStyles = _ref.innerStyles,
      barStyles = _ref.barStyles,
      dataStyles = _ref.dataStyles,
      _ref$dataInterpolatio = _ref.dataInterpolation,
      dataInterpolation = _ref$dataInterpolatio === void 0 ? function (p) {
    return "Loading " + p.toFixed(2) + "%";
  } : _ref$dataInterpolatio,
      _ref$initialState = _ref.initialState,
      initialState = _ref$initialState === void 0 ? function (active) {
    return active;
  } : _ref$initialState;

  var _useProgress = loaders_useProgress_cjs.useProgress(),
      active = _useProgress.active,
      progress = _useProgress.progress;

  var transition = web.useTransition(initialState(active), {
    from: {
      opacity: 1,
      progress: 0
    },
    leave: {
      opacity: 0
    },
    update: {
      progress: progress / 100
    }
  });
  return transition(function (_ref2, active) {
    var progress = _ref2.progress,
        opacity = _ref2.opacity;
    return active && /*#__PURE__*/React__default['default'].createElement(web.a.div, {
      style: _extends__default['default']({}, styles.container, {
        opacity: opacity
      }, containerStyles)
    }, /*#__PURE__*/React__default['default'].createElement("div", null, /*#__PURE__*/React__default['default'].createElement("div", {
      style: _extends__default['default']({}, styles.inner, innerStyles)
    }, /*#__PURE__*/React__default['default'].createElement(web.a.div, {
      style: _extends__default['default']({}, styles.bar, {
        scaleX: progress
      }, barStyles)
    }), /*#__PURE__*/React__default['default'].createElement(web.a.span, {
      style: _extends__default['default']({}, styles.data, dataStyles)
    }, progress.to(dataInterpolation)))));
  });
}

exports.Loader = Loader;
