// drei/loaders/useProgress.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var THREE = require('three');
var create = require('zustand');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var create__default = /*#__PURE__*/_interopDefaultLegacy(create);

var useProgress = create__default['default'](function (set) {
  THREE.DefaultLoadingManager.onStart = function (item, loaded, total) {
    return set({
      active: true,
      item: item,
      loaded: loaded,
      total: total,
      progress: loaded / total * 100
    });
  };

  THREE.DefaultLoadingManager.onLoad = function () {
    return set({
      active: false
    });
  };

  THREE.DefaultLoadingManager.onError = function (item) {
    return set(function (state) {
      return {
        errors: [].concat(state.errors, [item])
      };
    });
  };

  THREE.DefaultLoadingManager.onProgress = function (item, loaded, total) {
    return set({
      item: item,
      loaded: loaded,
      total: total,
      progress: loaded / total * 100
    });
  };

  return {
    errors: [],
    active: false,
    progress: 0,
    item: '',
    loaded: 0,
    total: 0
  };
});

exports.useProgress = useProgress;
