// drei/shapes/index.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

require('@babel/runtime/helpers/extends');
require('@babel/runtime/helpers/objectWithoutPropertiesLoose');
require('react');
require('react-three-fiber');
var shapes_generated_cjs = require('./generated.cjs.js');
require('three');
var shapes_RoundedBox_cjs = require('./RoundedBox.cjs.js');



exports.Box = shapes_generated_cjs.Box;
exports.Circle = shapes_generated_cjs.Circle;
exports.Cone = shapes_generated_cjs.Cone;
exports.Cylinder = shapes_generated_cjs.Cylinder;
exports.Dodecahedron = shapes_generated_cjs.Dodecahedron;
exports.Extrude = shapes_generated_cjs.Extrude;
exports.Icosahedron = shapes_generated_cjs.Icosahedron;
exports.Lathe = shapes_generated_cjs.Lathe;
exports.Octahedron = shapes_generated_cjs.Octahedron;
exports.Parametric = shapes_generated_cjs.Parametric;
exports.Plane = shapes_generated_cjs.Plane;
exports.Polyhedron = shapes_generated_cjs.Polyhedron;
exports.Ring = shapes_generated_cjs.Ring;
exports.Sphere = shapes_generated_cjs.Sphere;
exports.Tetrahedron = shapes_generated_cjs.Tetrahedron;
exports.Torus = shapes_generated_cjs.Torus;
exports.TorusKnot = shapes_generated_cjs.TorusKnot;
exports.Tube = shapes_generated_cjs.Tube;
exports.RoundedBox = shapes_RoundedBox_cjs.RoundedBox;
