// drei/shapes/RoundedBox.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _extends = require('@babel/runtime/helpers/extends');
var _objectWithoutPropertiesLoose = require('@babel/runtime/helpers/objectWithoutPropertiesLoose');
var React = require('react');
var reactThreeFiber = require('react-three-fiber');
var THREE = require('three');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _extends__default = /*#__PURE__*/_interopDefaultLegacy(_extends);
var _objectWithoutPropertiesLoose__default = /*#__PURE__*/_interopDefaultLegacy(_objectWithoutPropertiesLoose);
var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

var eps = 0.00001;

function createShape(width, height, radius0) {
  var shape = new THREE.Shape();
  var radius = radius0 - eps;
  shape.absarc(eps, eps, eps, -Math.PI / 2, -Math.PI, true);
  shape.absarc(eps, height - radius * 2, eps, Math.PI, Math.PI / 2, true);
  shape.absarc(width - radius * 2, height - radius * 2, eps, Math.PI / 2, 0, true);
  shape.absarc(width - radius * 2, eps, eps, 0, -Math.PI / 2, true);
  return shape;
}

var RoundedBox = /*#__PURE__*/React.forwardRef(function RoundedBox(_ref, ref) {
  var _ref$args = _ref.args;
  _ref$args = _ref$args === void 0 ? [] : _ref$args;

  var _ref$args$ = _ref$args[0],
      width = _ref$args$ === void 0 ? 1 : _ref$args$,
      _ref$args$2 = _ref$args[1],
      height = _ref$args$2 === void 0 ? 1 : _ref$args$2,
      _ref$args$3 = _ref$args[2],
      depth = _ref$args$3 === void 0 ? 1 : _ref$args$3,
      _ref$radius = _ref.radius,
      radius = _ref$radius === void 0 ? 0.05 : _ref$radius,
      _ref$smoothness = _ref.smoothness,
      smoothness = _ref$smoothness === void 0 ? 4 : _ref$smoothness,
      children = _ref.children,
      rest = _objectWithoutPropertiesLoose__default['default'](_ref, ["args", "radius", "smoothness", "children"]);

  var shape = React.useMemo(function () {
    return createShape(width, height, radius);
  }, [width, height, radius]);
  var params = React.useMemo(function () {
    return {
      depth: depth - radius * 2,
      bevelEnabled: true,
      bevelSegments: smoothness * 2,
      steps: 1,
      bevelSize: radius - eps,
      bevelThickness: radius,
      curveSegments: smoothness
    };
  }, [depth, radius, smoothness]);
  var geomRef = reactThreeFiber.useUpdate(function (geometry) {
    return void geometry.center();
  }, [shape, params]);
  return /*#__PURE__*/React__default['default'].createElement("mesh", _extends__default['default']({
    ref: ref
  }, rest), /*#__PURE__*/React__default['default'].createElement("extrudeBufferGeometry", {
    attach: "geometry",
    ref: geomRef,
    args: [shape, params]
  }), children);
});

exports.RoundedBox = RoundedBox;
